File: tic-tac-toe/tic-tac-toe.py
import pygame
import sys
from typing import List, Optional, Union

# Monokai Theme Colors
BG_COLOR = (39, 40, 34)        # #272822
GRID_COLOR = (230, 219, 116)   # #E6DB74
X_COLOR = (249, 38, 114)       # #F92672
O_COLOR = (166, 226, 46)       # #A6E22E
TEXT_COLOR = (248, 248, 242)   # #F8F8F2

# Game Constants
WIDTH, HEIGHT = 600, 600
LINE_WIDTH = 10
SQUARE_SIZE = WIDTH // 3

class TicTacToe:
    def __init__(self):
        pygame.init()
        self.screen = pygame.display.set_mode((WIDTH, HEIGHT))
        pygame.display.set_caption("Tic-Tac-Toe (Monokai)")
        self.font = pygame.font.SysFont("Arial", 80, bold=True)
        self.msg_font = pygame.font.SysFont("Arial", 40, bold=True)
        self.board: List[List[Optional[str]]] = [[None for _ in range(3)] for _ in range(3)]
        self.current_player = "X"
        self.game_over = False
        self.winner: Optional[str] = None

    def restart_game(self):
        self.board = [[None for _ in range(3)] for _ in range(3)]
        self.current_player = "X"
        self.game_over = False
        self.winner = None

    def draw_lines(self):
        # Vertical lines
        pygame.draw.line(self.screen, GRID_COLOR, (SQUARE_SIZE, 0), (SQUARE_SIZE, HEIGHT), LINE_WIDTH)
        pygame.draw.line(self.screen, GRID_COLOR, (SQUARE_SIZE * 2, 0), (SQUARE_SIZE * 2, HEIGHT), LINE_WIDTH)
        # Horizontal lines
        pygame.draw.line(self.screen, GRID_COLOR, (0, SQUARE_SIZE), (WIDTH, SQUARE_SIZE), LINE_WIDTH)
        pygame.draw.line(self.screen, GRID_COLOR, (0, SQUARE_SIZE * 2), (WIDTH, SQUARE_SIZE * 2), LINE_WIDTH)

    def draw_symbols(self):
        for row in range(3):
            for col in range(3):
                if self.board[row][col] == "X":
                    text = self.font.render("X", True, X_COLOR)
                    text_rect = text.get_rect(center=(col * SQUARE_SIZE + SQUARE_SIZE // 2, row * SQUARE_SIZE + SQUARE_SIZE // 2))
                    self.screen.blit(text, text_rect)
                elif self.board[row][col] == "O":
                    text = self.font.render("O", True, O_COLOR)
                    text_rect = text.get_rect(center=(col * SQUARE_SIZE + SQUARE_SIZE // 2, row * SQUARE_SIZE + SQUARE_SIZE // 2))
                    self.screen.blit(text, text_rect)

    def check_winner(self) -> Optional[str]:
        # Check rows and columns
        for i in range(3):
            if self.board[i][0] and self.board[i][0] == self.board[i][1] == self.board[i][2] and self.board[i][0] is not None:
                return self.board[i][0]
            if self.board[0][i] and self.board[0][i] == self.board[1][i] == self.board[2][i] and self.board[0][i] is not None:
                return self.board[0][i]
        
        # Check diagonals
        if self.board[0][0] and self.board[0][0] == self.board[1][1] == self.board[2][2] and self.board[0][0] is not None:
            return self.board[0][0]
        if self.board[0][2] and self.board[0][2] == self.board[1][1] == self.board[2][0] and self.board[0][2] is not None:
            return self.board[0][2]
        
        # Check draw
        if all(cell is not None for row in self.board for cell in row):
            return "Draw"
        
        return None

    def run(self):
        while True:
            self.screen.fill(BG_COLOR)
            self.draw_lines()
            self.draw_symbols()
            
            if self.game_over:
                if self.winner == "Draw":
                    msg = self.msg_font.render("Draw!", True, TEXT_COLOR)
                else:
                    msg = self.msg_font.render(f"{self.winner} Wins!", True, TEXT_COLOR)
                msg_rect = msg.get_rect(center=(WIDTH // 2, HEIGHT // 2))
                self.screen.blit(msg, msg_rect)
                
                restart_msg = self.msg_font.render("Press R to Restart", True, GRID_COLOR)
                restart_rect = restart_msg.get_rect(center=(WIDTH // 2, HEIGHT // 2 + 80))
                self.screen.blit(restart_msg, restart_rect)
                
                for event in pygame.event.get():
                    if event.type == pygame.QUIT:
                        pygame.quit()
                        sys.exit()
                    if event.type == pygame.KEYDOWN:
                        if event.key == pygame.K_r:
                            self.restart_game()
                            break
                
                pygame.display.update()
                continue

            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    sys.exit()
                
                if event.type == pygame.MOUSEBUTTONDOWN and not self.game_over:
                    mouseX = event.pos[0]
                    mouseY = event.pos[1]
                    row = mouseY // SQUARE_SIZE
                    col = mouseX // SQUARE_SIZE
                    
                    if self.board[row][col] is None:
                        # Player turn
                        self.board[row][col] = "X"
                        winner = self.check_winner()
                        if winner:
                            self.game_over = True
                            self.winner = winner
                        else:
                            # Bot turn
                            self.current_player = "O"
                            for r in range(3):
                                for c in range(3):
                                    if self.board[r][c] is None:
                                        self.board[r][c] = "O"
                                        winner = self.check_winner()
                                        if winner:
                                            self.game_over = True
                                            self.winner = winner
                                        else:
                                            self.current_player = "X"
                                        break
                                else: continue
                                break

            pygame.display.update()

if __name__ == "__main__":
    game = TicTacToe()
    game.run()

File: tic-tac-toe/tic-tac-toe.cpp
#include <SFML/Graphics.hpp>
#include <iostream>
#include <vector>
#include <string>

// Monokai Theme Colors
const sf::Color BG_COLOR(39, 40, 34);        // #272822
const sf::Color GRID_COLOR(230, 219, 116);   // #E6DB74
const sf::Color X_COLOR(249, 38, 114);       // #F92672
const sf::Color O_COLOR(166, 226, 46);       // #A6E22E
const sf::Color TEXT_COLOR(248, 248, 242);   // #F8F8F2

// Game Constants
const int WIDTH = 600;
const int HEIGHT = 600;
const float LINE_WIDTH = 10.0f;
const int SQUARE_SIZE = WIDTH / 3;

class TicTacToe {
private:
    sf::RenderWindow window;
    sf::Font font;
    std::vector<std::vector<std::string>> board;
    std::string current_player;
    bool game_over;
    std::string winner;

public:
    TicTacToe() :
        window(sf::VideoMode({(unsigned int)WIDTH, (unsigned int)HEIGHT}), "Tic-Tac-Toe (SFML 3)"),
        board(3, std::vector<std::string>(3, "")),
        current_player("X"),
        game_over(false),
        winner("")
    {
        window.setFramerateLimit(60);

        std::vector<std::string> system_fonts = {
            "/usr/share/fonts/TTF/DejaVuSans.ttf",
            "/usr/share/fonts/liberation/LiberationSans-Regular.ttf",
            "/usr/share/fonts/noto/NotoSans-Regular.ttf",
            "/usr/share/fonts/ttf-linux-libertine/LinLibertine_I.ttf",
            "Arial.ttf"
        };

        bool loaded = false;
        for (const auto& font_path : system_fonts) {
            if (font.openFromFile(font_path)) {
                std::cout << "Successfully loaded font: " << font_path << std::endl;
                loaded = true;
                break;
            }
        }

        if (!loaded) {
            std::cerr << "Error: Could not load any system fonts. Please place 'Arial.ttf' in the executable directory." << std::endl;
        }
    }

    void restart_game() {
        board.assign(3, std::vector<std::string>(3, ""));
        current_player = "X";
        game_over = false;
        winner = "";
    }

    void draw_lines() {
        sf::VertexArray lines(sf::PrimitiveType::Lines, 8);
        lines[0].position = { (float)SQUARE_SIZE, 0.f };
        lines[1].position = { (float)SQUARE_SIZE, (float)HEIGHT };
        lines[2].position = { (float)SQUARE_SIZE * 2, 0.f };
        lines[3].position = { (float)SQUARE_SIZE * 2, (float)HEIGHT };
        lines[4].position = { 0.f, (float)SQUARE_SIZE };
        lines[5].position = { (float)WIDTH, (float)SQUARE_SIZE };
        lines[6].position = { 0.f, (float)SQUARE_SIZE * 2 };
        lines[7].position = { (float)WIDTH, (float)SQUARE_SIZE * 2 };

        for (int i = 0; i < 8; ++i) lines[i].color = GRID_COLOR;
        window.draw(lines);
    }

    void draw_symbols() {
        for (int row = 0; row < 3; ++row) {
            for (int col = 0; col < 3; ++col) {
                if (board[row][col] == "X") {
                    sf::Text text(font, "X", 80);
                    text.setFillColor(X_COLOR);
                    text.setOrigin({40.f, 40.f});
                    text.setPosition({(float)col * SQUARE_SIZE + SQUARE_SIZE / 2.0f, (float)row * SQUARE_SIZE + SQUARE_SIZE / 2.0f});
                    window.draw(text);
                } else if (board[row][col] == "O") {
                    sf::Text text(font, "O", 80);
                    text.setFillColor(O_COLOR);
                    text.setOrigin({40.f, 40.f});
                    text.setPosition({(float)col * SQUARE_SIZE + SQUARE_SIZE / 2.0f, (float)row * SQUARE_SIZE + SQUARE_SIZE / 2.0f});
                    window.draw(text);
                }
            }
        }
    }

    std::string check_winner() {
        for (int i = 0; i < 3; ++i) {
            if (board[i][0] != "" && board[i][0] == board[i][1] && board[i][0] == board[i][2]) return board[i][0];
            if (board[0][i] != "" && board[0][i] == board[1][i] && board[0][i] == board[2][i]) return board[0][i];
        }
        if (board[0][0] != "" && board[0][0] == board[1][1] && board[0][0] == board[2][2]) return board[0][0];
        if (board[0][2] != "" && board[0][2] == board[1][1] && board[0][2] == board[2][0]) return board[0][2];

        bool draw = true;
        for (int i = 0; i < 3; ++i) {
            for (int j = 0; j < 3; ++j) {
                if (board[i][j] == "") draw = false;
            }
        }
        if (draw) return "Draw";

        return "";
    }

    void run() {
        while (window.isOpen()) {
            while (auto ev = window.pollEvent()) {
                if (ev->is<sf::Event::Closed>()) {
                    window.close();
                }
                if (ev->is<sf::Event::MouseButtonPressed>()) {
                    const auto* mouse_event = ev->getIf<sf::Event::MouseButtonPressed>();
                    if (mouse_event && mouse_event->button == sf::Mouse::Button::Left && !game_over) {
                        const auto& mouse_pos = mouse_event->position;
                        int row = mouse_pos.y / SQUARE_SIZE;
                        int col = mouse_pos.x / SQUARE_SIZE;

                        if (row >= 0 && row < 3 && col >= 0 && col < 3 && board[row][col] == "") {
                            board[row][col] = current_player;
                            std::string res = check_winner();
                            if (res != "") {
                                game_over = true;
                                winner = res;
                            } else {
                                current_player = (current_player == "X") ? "O" : "X";
                            }
                        }
                    }
                }
                if (ev->is<sf::Event::KeyPressed>()) {
                    if (ev->getIf<sf::Event::KeyPressed>()->code == sf::Keyboard::Key::R) {
                        restart_game();
                    }
                }
            }

            window.clear(BG_COLOR);
            draw_lines();
            draw_symbols();

            if (game_over) {
                sf::String win_text = (winner == "Draw") ? "Draw!" : (sf::String(winner) + " Wins!");
                sf::Text msg(font, win_text, 40);
                msg.setFillColor(TEXT_COLOR);
                msg.setPosition({WIDTH / 2.0f - msg.getGlobalBounds().size.x / 2.0f, HEIGHT / 2.0f - msg.getGlobalBounds().size.y / 2.0f});
                window.draw(msg);

                sf::Text restart_msg(font, "Press R to Restart", 30);
                restart_msg.setFillColor(GRID_COLOR);
                restart_msg.setPosition({WIDTH / 2.0f - restart_msg.getGlobalBounds().size.x / 2.0f, HEIGHT / 2.0f + 50});
                window.draw(restart_msg);
            }

            window.display();
        }
    }
};

int main() {
    TicTacToe game;
    game.run();
    return 0;
}
